Add `match_arm_indent` option

KAAtheWiseGit opened the pull request:
Allows to disable the indentation of match arms.

Related issue: #2937

ytmimi commented on src/matches.rs:
```diff
@@ -100,17 +100,22 @@ pub(crate) fn rewrite_match(
     let inner_attrs_str = if inner_attrs.is_empty() {
         String::new()
     } else {
         let shape = if context.config.style_edition() <= StyleEdition::Edition2021 {
             shape
+        } else if !context.config.match_arm_indent() {
+            shape
```
Instead of adding another branch, why not just add an `or` condition to the preceding check?

```rust
let shape = if context.config.style_edition() <= StyleEdition::Edition2021
    || !context.config.match_arm_indent() 
{
    shape
} else {
    shape.block_indent(context.config.tab_spaces())
};
```

KAAtheWiseGit replied:
I think I was worried about the condition being too long, but yeah, this looks more readable, I'll fold it

ytmimi commented on tests/source/match-unindent.rs:
let's move all of these `source` test cases to a new `tests/source/configs/match_arm_indent` folder.

ytmimi commented on tests/source/match-unindent.rs:
```diff
@@ -0,0 +1,19 @@
+
+fn foo() {
+    match x {
+        a => {
+            "line1";
+            "line2"
+        }
+        ThisIsA::Guard if true => {
+            "line1";
+            "line2"
+        }
+        ThisIsA::ReallyLongPattern(ThatWillForce::TheGuard, ToWrapOnto::TheFollowingLine) if true => {
+            "line1";
+            "line2"
+        }
```
Let's expand on the test cases we've got here. It would be nice to add a match arm that has an attribute and one that has a long pattern that wraps to the next line, maybe using a long `OR` pattern and one where the match arm guard is on the next line.

KAAtheWiseGit replied:
Huh, I didn't know that match arm attributes were a thing.  Added a separate test file for that

ytmimi commented on tests/target/match-noindent-tab.rs:
let's move all of these `target` test cases to a new `tests/target/configs/match_arm_indent` folder.

ytmimi replied:
Might also be nice to add a test case to show how `match_arm_indent` interacts with `match_arm_leading_pipes`.

ytmimi commented on Configurations.md:
```diff
@@ -1862,6 +1862,42 @@ fn foo() {
 
+## `match_arm_indent`
+
+Controls whether match arms are indented.  If disabled, the match patterns will be located on the indentation level as the match statement itself.  This means match blocks will only be indented once, not twice.
```
Some minor tweaks to the wording. Feel free to edit what I'm suggesting.

```suggestion
Controls whether match arms are indented.  If disabled, match arms will be formatted at the same indentation level as the outer `match` statement.  Meaning that match blocks will only be indented once, not twice.
```

ytmimi reviewed (changes requested):
Thanks for taking the time to work on this older feature.

The code seems pretty straightforward to me, but I've left some comments inline. Take a look when you've got a chance.

KAAtheWiseGit commented on Configurations.md:
```diff
@@ -2475,7 +2511,7 @@ required_version=">=1.0.0, <2.0.0"
 
 The width threshold for an array element to be considered "short".
 
-The layout of an array is dependent on the length of each of its elements. 
+The layout of an array is dependent on the length of each of its elements.
```
Oops, missed that one in the diff

KAAtheWiseGit commented:
Applied the changes, moved the tests to the `configs/match_arm_indent` folder, and added a bunch more tests. I also remembered that this feature interacts with nested matches, so I added a few cases for that

ytmimi commented:
> Applied the changes, moved the tests to the configs/match_arm_indent folder, and added a bunch more tests. I also remembered that this feature interacts with nested matches, so I added a few cases for that

Thanks for making the changes and expanding on the test case. I'll have some time to re-review this early next week.

ytmimi commented on tests/target/configs/match_arm_indent/nested.rs:
```diff
@@ -0,0 +1,79 @@
+    }));
+}
+
+fn silly() {
+    match value {
+    Inner(i1) => match i1 {
+    Inner(i2) => match i2 {
+    Inner(i3) => match i3 {
+    Inner(i4) => match i4 {
+    Inner => "it's a readability tradeoff, really",
+    },
+    },
+    },
+    },
+    }
+}
```
Good test case, but it's pretty unfortunate formatting. It's really hard to tell that these are nested match statements.

KAAtheWiseGit replied:
Yeah, that's the main problem with this formatting option.  The example is deliberately obtuse, but I think it's somewhat better with realistic usage.  That's how I envision it working for AST matching, for example:

```rust
match statement {
Stmt::Local(item) => match item {
Item::Const(item_const) => {
    // do stuff with const
}
Item::Enum(item_enum) => {
    // do stuff with enum
}
// rets of items
}
Stmt::Expr(expr, semi) => match expr {
Expr::Array(arr) => {
     // do stuff with arr (and semi, with is in scope despite not being in the second match statement)
}
Expr::Assign(assign) => {
    //  ...
}
// ...
}
}
```

Here the flat structure is desirable, as we're processing all of the items together.

In theory, there could be a heuristic which would wrap the nested match as a body, indenting it.  But I'm not sure how it could be applied so that it both fixes these edge cases, but doesn't break sane nesting

KAAtheWiseGit commented:
Added the tracking issue

ytmimi reviewed (approved):
Overall the implementation is pretty straightforward and doesn't complicate match arm formatting. I still think the nested cases are hard to read, but I'd say we're fine to move forward with this as an unstable option.

ytmimi commented:
Just an FYI the new option won't be available on nightly until the next subtree sync with rust-lang/rust. hoping to get that out soonish
